File: src/mk2/PythonPackage/src/tamaf/wakerbehavior.py
import time
from abc import abstractmethod
from .behavior import Behavior

class WakerBehavior(Behavior):
    def __init__(self, agent, timeout_ms: int):
        super().__init__(agent)
        self.wake_up_time = time.time() + (timeout_ms / 1000.0)
        self.done_flag = False

    def Action(self):
        if time.time() >= self.wake_up_time:
            self.OnWake()
            self.done_flag = True

    @abstractmethod
    def OnWake(self):
        pass

    def Done(self):
        return self.done_flag

File: src/mk2/JavaPackage/src/main/java/tamaf/WakerBehavior.java
package tamaf;

public abstract class WakerBehavior extends Behavior {
    private long wakeUpTime;
    private boolean finished = false;

    public WakerBehavior(long timeoutMs) {
        super();
        this.wakeUpTime = System.currentTimeMillis() + timeoutMs;
    }

    public WakerBehavior(Agent agent, long timeoutMs) {
        super(agent);
        this.wakeUpTime = System.currentTimeMillis() + timeoutMs;
    }

    @Override
    public void action() {
        if (System.currentTimeMillis() >= wakeUpTime) {
            onWake();
            finished = true;
        }
    }

    public abstract void onWake();

    @Override
    public boolean done() {
        return finished;
    }
}
